plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/bitbucketServer.ts

```diff
@@ -237,10 +241,15 @@ export function createPublishBitbucketServerAction(options: {
         email: config.getOptionalString('scaffolder.defaultAuthor.email'),
       };
 
-      const auth = {
-        username: 'x-token-auth',
-        password: token,
-      };
+      const auth = authConfig.token
+        ? {
+            username: 'x-token-auth',
```

Comment:
Haha exactly... that's a bit confusing, but it works 😂 It's also mentioned here https://confluence.atlassian.com/bitbucketserver/personal-access-tokens-939515499.html
> HTTP access tokens in Bitbucket Data Center can be created for users as well as for teams working in projects and repositories. Use them in place of passwords for Git over HTTPS, or to authenticate when using the [Bitbucket REST API](https://developer.atlassian.com/server/bitbucket/how-tos/command-line-rest/).

> For Git operations, you can use your user's access token as a substitute for your password.

Reply:
I think we want both, however I think you can specify the token as the password with username and it will work as well, though it's not that intuitive.

So I would say we have 2 options:
1. Parse username when specifying token.
2. Only use username/password and document that token can be specified as password (then we can remove the token field as well). I think this is how it was used in the previous iteration.

Reply:
Looking at https://confluence.atlassian.com/bitbucketserver/personal-access-tokens-939515499.html again, it seems you could use username + token for Basic Auth (basically, you can use the token as password, too).

> For project or repository tokens, we recommend only using Bearer Auth without your username

But this would be token only, no username.

Example on how to use this at Git operations:

> Bearer Auth example for HTTP Git operations:
>
> ```git clone -c http.extraHeader='Authorization: Bearer MDM0MjM5NDc2MDxxxxxxxxxxxxxxxxxxxxx' https://bitbucketserver.com/scm/projectname/teamsinspace.git```

which is basically your 2nd option.

Otherwise, it would be more like Basic Auth, using the token as password

> For Git operations, you can use your user's access token as a substitute for your password. For example, to clone using an access token you can enter:

Overall, it means you *could* use the username + password with the token used as password and it would work for all use cases -- but it would use Basic Auth everywhere.

`x-token-auth` seems was never supported for it. Likely came from Bitbucket Cloud where you can use this as username when you use tokens.

And overall, it seems like there is not always a username available for a token (project or repo tokens). These would not be usable if a username will be required.

Reply:
```
/*
provider    username         password
GitHub      'x-access-token' token
BitBucket   'x-token-auth'   token
GitLab      'oauth2'         token
From : https://isomorphic-git.org/docs/en/onAuth with fix for GitHub

Azure       'notempty'      token
*/
```
from `packages/backend-common/src/scm/git.ts`

The example works for Bitbucket Cloud, not Bitbucket Server.

There is a headers option which currently only sets the user-agent header.

I assume it would be possible to add support for Bearer Auth. Without this, using username+password (or token as password) would be still an alternative. But then we can clean up the token option, here.

Reply:
Ah right, didn't think of the repository/project tokens.
But then I think it makes sense to keep username/password combination with possibility to substitute with token, as it is line with Bitbuckets documentation as well.
And repository/project tokens simply use the token field.
I verified both username/password & username/token in the `username`, `password` fields, both seem to work, (with the BitbucketDiscoveryProcessor as earlier, do we need to verify the actions @namco1992?).

Reply:
Hi @ONordander,

Yes I believe they should all work as expected with the `BitbucketDiscoveryProcessor`, because only REST API is used in the processor, the problematic part is the git operations.

I verified both username/password & username/token, and they should work with git. However, currently I don't have a pair of username/password with write access, so I verified read actions like `git clone` and it works for username/password. For username/token, both read and write work.

Reply:
I think the tricky part now is how do we implement this clearly without much confusion, as the currently different approaches of auth is a bit confusing already.

I think what @ONordander proposed works for me:
> Only use username/password and document that token can be specified as password (then we can remove the token field as well).

Then we can add support for extra headers in `initRepoAndPush` (not sure if there is anywhere else) to address the scenario that only token is provided.

Reply:
If the proposal of removing works for you, we would not need to add support for extra headers, etc. Would we need to support the token only scenario?

Reply:
Here is a quick draft of how support for Bearer auth could look like: PR #12785

Reply:
TBH I don't see the necessity of supporting the token-only scenario, especially in the scaffolder actions. Besides, it would incur additional changes like your new PR. Maybe we can consider adding the support if other users in the future require it?

Reply:
Sounds good to me.

@ONordander I assume you are ok with this, too.

@dhruvaldarji @awanlin as Bitbucket Server users as far as I remember: Do you have any token-only use case?

Reply:
Thanks for including me on this @pjungermann. 

**TL;DR;** We use a combination of just the token in some cases and username and token in others

We are on older version of Bitbucket server 7.2.4. We use static locations (this just uses the token) for the catalog and then have two custom plugins that interact with our Bitbucket. One uses the UrlReader (this just uses the token) and the other clones the repo to get stats from the Git history. The second one we are using a username and token combination, that was the only thing that worked. I spent hours trying to get `x-token-auth` to work because I've found the Atlassian docs terribly difficult to use as an outsider to their products.

Reply:
I've added two more commits. The first is the completed change from the draft PR #12785 and made use of it in another commit.

This should allow now

- (username, password)
- (username, token)
- (token)

for all the use cases.

Would be great if you can test it again.

Reply:
Verified the latest with the BitbucketDiscoveryProcessor and the following cases all seem to work:
- username/password
- username/token
- token

Nice work :)

Reply:
Thanks for testing the change!

@ONordander Would be great if you can leave a code review, too, to let @benjdlambert know that all is good from your end. This would make it easier for him. Same for you, @namco1992 (esp. as it will resolve the bug reported by you).

Reply:
Looks good from my end, just a minor copy-paste error :)
Looking forward to get this merged

Reply:
This PR will save us here in our company
thank you so much